core/templates/domain/statistics/PlaythroughIssueObjectFactory.ts

```diff
@@ -20,13 +20,36 @@
 import { Injectable } from '@angular/core';
 import { downgradeInjectable } from '@angular/upgrade/static';
 
+export interface IMultipleIncorrectSubmissionsCustomizationArgs {
+  'state_name': {value: string};
+  'num_times_answered_incorrectly': {value: number};
+}
+
+export interface ICyclicStateTransitionsCustomizationArgs {
+  'state_names': {value: string[]};
+}
+
+export interface IEarlyQuitCustomizationArgs {
+  'state_name': {value: string};
+  'time_spent_in_exp_in_msecs': {value: number};
+}
```

Comment:
Maybe use the discriminated unions?

Reply:
I tried doing it but here the problem was that the discriminating factor i.e. issue_type was itself a member of the class. So, the problem really arises when we have to assign type to `issueType` and `issueCustomizationArgs` in the class. 

We can assign `issueType` as `EarlyQuit | MultipleChoice .....` and `issueCustomizationArgs` as `IEarlyQuitCustomizationArgs | ICyclicStateTransitionsCustomizationArgs ....` but it is sematically wrong because the compiler doesn't know `EarlyQuit` can only be present with `IEarlyQuitCustomizationArgs` and so on. So, it creates errors.

/cc @brianrodri

Reply:
So I've been spending this morning giving it a try, and I came up with the following code:

```typescript
export interface IEarlyQuitCustomizationArgs {
  'state_name': {value: string};
  'time_spent_in_exp_in_msecs': {value: number};
}

export interface IMultipleIncorrectSubmissionsCustomizationArgs {
  'state_name': {value: string};
  'num_times_answered_incorrectly': {value: number};
}

export interface ICyclicStateTransitionsCustomizationArgs {
  'state_names': {value: string[]};
}

// NOTE TO DEVELOPERS: Treat this as an implementation detail; do not export it.
interface IPlaythroughBackendDictBase<ICustomizationArgs> {
  'issue_customization_args': ICustomizationArgs;
  'playthrough_ids': string[];
  'schema_version': number;
  'is_valid': boolean;
}

// NOTE TO DEVELOPERS: Treat this as an implementation detail; do not export it.
class PlaythroughIssueBase<ICustomizationArgs> {
  constructor(
      public issueCustomizationArgs: ICustomizationArgs,
      public playthroughIds: string[],
      public schemaVersion: number,
      public isValid: boolean) {}

  toBackendDict(): IPlaythroughBackendDictBase<ICustomizationArgs> {
    return {
      issue_customization_args: this.issueCustomizationArgs,
      playthrough_ids: this.playthroughIds,
      schema_version: this.schemaVersion,
      is_valid: this.isValid,
    };
  }
}

export interface IEarlyQuitPlaythroughIssueBackendDict extends
    IPlaythroughBackendDictBase<IEarlyQuitCustomizationArgs> {
  'issue_type': 'EarlyQuit';
}

export interface IMultipleIncorrectSubmissionsPlaythroughIssueBackendDict
    extends IPlaythroughBackendDictBase<
      IMultipleIncorrectSubmissionsCustomizationArgs>{
  'issue_type': 'MultipleIncorrectSubmissions';
}

export interface ICyclicStateTransitionsPlaythroughIssueBackendDict
    extends IPlaythroughBackendDictBase<
      ICyclicStateTransitionsCustomizationArgs>{
  'issue_type': 'CyclicStateTransitions';
}

export type IPlaythroughIssueBackendDict = (
  ICyclicStateTransitionsPlaythroughIssueBackendDict |
  IEarlyQuitPlaythroughIssueBackendDict |
  IMultipleIncorrectSubmissionsPlaythroughIssueBackendDict);

export class EarlyQuitPlaythroughIssue extends
    PlaythroughIssueBase<IEarlyQuitCustomizationArgs> {
  public readonly issueType: 'EarlyQuit' = 'EarlyQuit';
}

export class MultipleIncorrectSubmissionsPlaythroughIssue extends
    PlaythroughIssueBase<IMultipleIncorrectSubmissionsCustomizationArgs> {
  public readonly issueType: 'EarlyQuit' = 'EarlyQuit';
}

export class CyclicStateTransitionsPlaythroughIssue extends
    PlaythroughIssueBase<ICyclicStateTransitionsCustomizationArgs> {
  public readonly issueType: 'EarlyQuit' = 'EarlyQuit';
}

export type PlaythroughIssue = (
  EarlyQuitPlaythroughIssue | MultipleIncorrectSubmissionsPlaythroughIssue |
  CyclicStateTransitionsPlaythroughIssue);

@Injectable({
  providedIn: 'root'
})
export class PlaythroughIssueObjectFactory {
  createFromBackendDict(
      backendDict: IPlaythroughIssueBackendDict): PlaythroughIssue {
    switch (backendDict.issue_type) {
      case 'EarlyQuit':
        return new EarlyQuitPlaythroughIssue(
          backendDict.issue_customization_args, backendDict.playthrough_ids,
          backendDict.schema_version, backendDict.is_valid);
      case 'MultipleIncorrectSubmissions':
        return new MultipleIncorrectSubmissionsPlaythroughIssue(
          backendDict.issue_customization_args, backendDict.playthrough_ids,
          backendDict.schema_version, backendDict.is_valid);
      case 'CyclicStateTransitions':
        return new CyclicStateTransitionsPlaythroughIssue(
          backendDict.issue_customization_args, backendDict.playthrough_ids,
          backendDict.schema_version, backendDict.is_valid);
      default:
        const invalidBackendDict: never = backendDict;
        throw new Error(
          'Backend dict does not match any known issue type: ' +
          invalidBackendDict);
    }
  }
}
```

Note that this will almost certainly result in compilation errors around the code base, as it should ;)
However, the error messages printed by the compilation errors will help you fix them (and demonstrate how helpful these types can be!!)

Reply:
@nishantwrp FYI I've updated the code a lot after looking through the compilation errors myself, this newer version should provide all of the necessary debug information (and should be our template for new discriminated unions moving forward). PTAL!

Reply:
@nishantwrp FYI I've updated the code a lot after looking through the compilation errors myself, this newer version should provide all of the necessary debug information (and should be our template for new discriminated unions moving forward).

PTAL!

/cc @vojtechjelinek

Reply:
Done! I really liked your approach @brianrodri and yeah, this can be used as a template for this kind of stuff.

Reply:
Awesome, happy to hear it :)

Reply:
Hi @brianrodri, I've refactored your code a bit. Please see if it looks good! (See my latest commit)